todo_app/app.py

```diff
@@ -3,26 +3,99 @@
 import os
 import requests
 from todo_app.mongo_actions import MongoActions
+from flask_login import LoginManager, login_required, login_user, UserMixin, current_user
+from oauthlib.oauth2 import WebApplicationClient
+from functools import wraps
+
+class User(UserMixin):
+    def __init__(self, userId):
+        self.id = userId
+        if userId == 'jainsushma':
+            self.role = "writer"
+        else: 
+            self.role = "reader"
 
 def create_app():
     app = Flask(__name__)
+    app.config.from_object('todo_app.flask_config.Config')
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    # Values setup during registration
+    clientId = os.getenv('CLIENTID')
+    clientSecret = os.getenv('CLIENTSECRET')
+    loginDisabled = app.config.get('LOGIN_DISABLED')
+    githubClient = WebApplicationClient(clientId)
+
+    @app.route('/login/callback')
+    def call_back():
+        callbackCode = request.args.get("code")
+        tokenUrl, headers, body = githubClient.prepare_token_request("https://github.com/login/oauth/access_token", code=callbackCode)
+        headers["Accept"] = "application/json"
+        r = requests.post(
+            tokenUrl,
+            headers=headers,
+            data=body,
+            auth=(clientId, clientSecret),
+        )
+        params=githubClient.parse_request_body_response(r.text)
+        print(params['access_token'])
+        tokenUrl, headers, body = githubClient.add_token("https://api.github.com/user")
+        userInfo = requests.get(
+            tokenUrl,
+            headers=headers,
+            data=body,
+        )
+        userInfo = userInfo.json()
+        print((userInfo['login']))
+        login_user(load_user(userInfo['login']))
+        return redirect('/')
+
+    @login_manager.unauthorized_handler 
+    def unauthenticated():
+        githubRedirect = githubClient.prepare_request_uri("https://github.com/login/oauth/authorize")
+        print(githubRedirect)
+        return redirect(githubRedirect)
+
+    @login_manager.user_loader 
+    def load_user(user_id):
+        user = User(user_id)
+        return user
+
+    def validate_user_role(func):
+        @wraps(func)
+        def wrapTheFunction(*args, **kwargs):
+            if loginDisabled or (current_user.role == "reader"):
+                if loginDisabled:
+                    return func(*args, **kwargs)
+                elif (current_user.role == "reader"):    
+                    return render_template("error.html", error="Insufficient User Rights") 
+                return func(*args, **kwargs)    
```

Comment:
I think you've got a mismatch here - if security is enabled and the role is "writer" then the user doesn't trigger the outer if, and nothing happens - so the wrapped function doesn't execute at all (which leads to errors)

Removing the outer if statement doesn't change most cases, but does fix the error:
```suggestion
            if loginDisabled:
                return func(*args, **kwargs)
            elif (current_user.role == "reader"):    
                return render_template("error.html", error="Insufficient User Rights") 
            return func(*args, **kwargs)    
```

Having said that, note that loginDisabled treats the user as a writer, so we could simplify further to:
```suggestion
            if (current_user.role == "reader"):    
                return render_template("error.html", error="Insufficient User Rights") 
            return func(*args, **kwargs)    
```

Reply:
Done

Reply:
It seems need to make use of loginDisabled to keep e2e tests passing in Github actions build. Hope following changes are ok
`  if (not loginDisabled and current_user.role == "reader"):    
                return render_template("error.html", error="Insufficient User Rights") 
            return func(*args, **kwargs) `

Reply:
Ah that makes sense - `current_user` isn't set if login is disabled. That sounds reasonable to me 👍